Initial buildout

pjhartzell opened the pull request:
**Description:**
Initial build of project after boilerplate creation.

**PR checklist:**

- [x] Code is formatted (run `scripts/format`).
- [x] Code lints properly (run `scripts/lint`).
- [x] Tests pass (run `scripts/test`).
- [x] Documentation has been updated to reflect changes, if applicable.
- [x] Examples have been updated to reflect changes, if applicable
- [ ] Changes are added to the [CHANGELOG](../CHANGELOG.md).

pjhartzell committed:
progress push

pjhartzell committed:
creating items

pjhartzell committed:
get tests running

pjhartzell committed:
a pin before implementing some large changes

pjhartzell committed:
condense fragments into single assets and collection file

pjhartzell committed:
backlog of improvements

pjhartzell committed:
get collection command running; add docstrings to all functions

pjhartzell committed:
add HLS product to Items; update readme

gadomski commented on README.md:
```diff
@@ -11,26 +11,47 @@
+  - [view](https://github.com/stac-extensions/view)
 - Extra fields:
-  - `hls:tile-id`: Tile identifier using the Military Grid Reference System (MGRS)
+  - `hls:product`: "HLSL30" (Landsat) or "HLSS30" (Sentinel) product
```
What are your thoughts on implementing a custom STAC extension, ala https://github.com/stactools-packages/noaa-mrms-qpe/tree/main/extension? Too much work for too little payoff?

pjhartzell replied:
First reaction is, yes, it seems a bit much for a single field. Not that it wouldn't be a good exercise. 

Add as a `feature request` issue?

pjhartzell replied:
See #2

gadomski commented on src/stactools/hls/commands.py:
```diff
@@ -13,43 +17,119 @@ def create_hls_command(cli: Group) -> Command:
-
+    @click.option(
+        "-a",
+        "--antimeridian_strategy",
```
```suggestion
        "--antimeridian-strategy",
```

gadomski commented on src/stactools/hls/commands.py:
```diff
@@ -13,43 +17,119 @@ def create_hls_command(cli: Group) -> Command:
+        collection.update_extent_from_items()
+
+        collection.catalog_type = CatalogType.SELF_CONTAINED
+        collection.make_all_asset_hrefs_relative
```
```suggestion
        collection.make_all_asset_hrefs_relative()
```

pjhartzell replied:
Erf. Done.

gadomski commented on src/stactools/hls/fragments/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    """Class for accessing collection and asset data."""
+
+    def __init__(self) -> None:
+        self.assets = self._load("assets.json")
+        for key in self.assets.keys():
+            self.assets[key]["type"] = MediaType.COG
```
Is `assets` meant to be accessed directly, or only through the `self.asset()` method? If it is not supposed to be used directly, this attribute should be made private (`self._assets`).

pjhartzell replied:
It is accessed [directly in `create_collection`](https://github.com/stactools-packages/hls/blob/82dfa755f6a833dcb15107579346f73a4bb3496d/src/stactools/hls/stac.py#L135) for building `item_assets`.

gadomski commented on src/stactools/hls/stac.py:
```diff
@@ -1,117 +1,154 @@
-    See `Collection<https://pystac.readthedocs.io/en/latest/api.html#collection>`_.
+    Args:
+        cog_href (str): HREF to one of the EO COG files in the granule.
+        read_href_modifier (Optional[ReadHrefModifier], optional): An optional
```
```suggestion
        read_href_modifier (ReadHrefModifier, optional): An optional
```

pjhartzell replied:
Thanks. Cleaned up all other occurrences as well.

pjhartzell committed:
fix: change antimeridian_strategy command option underscore to a hyphen

pjhartzell committed:
fix: add missing parentheses to `make_all_asset_hrefs_relative` function call

pjhartzell committed:
fix: remove occurrences of redundant `Optional` from docstrings

sharkinsspatial commented:
@pjhartzell I just stumbled on this repo while looking up another stactools repo 🎊 .  We currently generate inline STAC items for distribution with the HLS data as part of the processing pipeline using this older, poorly maintained lib https://github.com/NASA-IMPACT/hls-cmr_stac.

Not sure what your ingestion plans are for the HLS data but we can definitely consider adopting this repo for generating items in our next HLS pipeline minor release if that would streamline things for you and provide consistent items (Prior to implementing the `projection` extension we avoided directly accessing files via `rasterio` but  that is a moot point now).

gadomski commented:
@sharkinsspatial, @pjhartzell's on PTO for the week so I'll jump in just to say we'd love to collaborate here! Our plans are to ingest HLS into Microsoft Planetary Computer, though we intentionally keep all Microsoft-specific data/logic out of this repo.

If there were STAC metatada provided w/ HLS we would love to use it, e.g. we use USGS's Landsat STAC when possible. So much the better if its all coming from the same repo.

sharkinsspatial commented:
@gadomski I'll try to get some details on when our next minor pipeline release will be but in the near term, once this PR is merged I'll start testing this package in our development environment. 

A few notes about HLS data consistency that might be helpful for your ingestion/usage.  

- All data produced prior to mid-May 2022 uses a CRS with incomplete datum information due to an issue with the underlying HDF-EOS lib used in our pipeline.  This CRS information was updated in our production system in mid-May with https://github.com/NASA-IMPACT/hls-hdf_to_cog/pull/9/files.  With older versions of `rasterio` this incomplete datum information can result in range read performance issues due to some EPSG code lookup paths.  This performance issue is fixed with https://github.com/rasterio/rasterio/releases/tag/1.3b3 but there is discussion of possible NASA reprocessing effort to update the CRS information for older data in the archive as well.

- The Landsat 9 data was included in the pipeline in our mid-May release.  The reverse processing for the Landsat 9 data from Sep 2021 through May 17 2022 will be released in mid-August.

- All the STAC items prior to mid-May have the same hemisphere EPSG issue as the Landsat stactools package https://github.com/stactools-packages/landsat/issues/12

pjhartzell commented on src/stactools/hls/metadata.py:
```diff
@@ -0,0 +1,148 @@
+        Returns:
+            Metadata: A Metadata dataclass
+        """
+        read_h5_href = utils.modify_href(cog_href, read_href_modifier)
+        with rasterio.open(read_h5_href) as dataset:
```
This `read_h5_href` term is a copy-paste leftover. Edit to remove the h5 term, .e.g., make it `read_cog_href`.

pjhartzell committed:
fix: correct a copy-paste leftover

pjhartzell committed:
fix: handle older COG CRS information

Prior to May 2022, the COG CRS WKT does not convert cleanly to an EPSG
code via rasterio's `crs.to_epsg()`. Now pulling the UTM zone from the
WKT and converting to the correct UTM (north) EPSG code.

pjhartzell commented:
@sharkinsspatial Good info, definitely interested in working with you on this stactools package!

- I updated the EPSG code extraction to handle the pre-May 2022 data. The update also forces the hemisphere to be north.
- Yep, we are handling Landsat 9 data. Right now the package will handle a combination of Landsat 8 and 9 data in a single granule. But that seems unlikely. Can you confirm/deny that granules are exclusive to a single platform? 
- The XML geometries are much better than the current use of a bounding box obtained from one of the COG image bounds (#3), so that will need to be updated. I was trying to limit the number of files touched, but worth opening the XML for the geometry data.
- A note about `eo:bands` in the [hls-cmr_stac](https://github.com/NASA-IMPACT/hls-cmr_stac) repo: I see instrument specific center wavelengths and FWHM values. My understanding is that the Sentinel data is bandpassed to match the Landsat spectral response, so we are using a single set of center wavelengths and FWHM values. Is my understanding correct?
